feat(menu): add nested menu functionality

crisbeto opened the pull request:
Adds the ability for an `md-menu-item` inside of a `md-menu` to trigger another `md-menu`. This is a first step towards a `md-toolbar` component.

Demo: https://material-cb7ec.firebaseapp.com/menu

Fixes #1429.

m98 commented:
1.  Would not be good if we could show a simple arrow (to right) for nested menus? Like user should not guess that does this menu item has nested item or not, and do this continuously for the whole list items.
![screenshot from 2017-07-05 13-43-43](https://user-images.githubusercontent.com/7496103/27857778-36ac6bf2-6188-11e7-8fe6-1480d41a12e6.png)

2. The animation seems to run very fast after mouse-in, a short delay would make it feel better for users, it's not good to animate too fast (I mean on-mouse-in on list items) because user is just mouse hovering the menu to find the item he is looking for, and the fast opening menu (without delay) is disturbing.
([See this short video](https://storage.googleapis.com/material-design/publish/material_v_11/assets/0B08MbvYZK1iNWVBad29DaVJrREk/components-menus-menus-cascading_dropdown_spec_large_xhdpi.mp4))

joejordanbrown commented:
We need to make sure that when using a mouse and running the cursor diagnally that it doesn't change to another nested menu item.

I haven't seen this mentioned in the design document, so I'm not sure if this has been overlooked at the moment.

The google documents menu does this but i beleive its done using a timer instead of detecting direction, a good example would be the amazon menu even if you move slowly it still has the correct nested menu open and moving the cursor up and down the menu shows the nested items without delay. This seems to be a lot more user friendly.

![image](https://user-images.githubusercontent.com/10200431/27863625-ec5b546e-6182-11e7-937a-3f5fd436d978.png)

This page [here](http://bjk5.com/post/44698559168/breaking-down-amazons-mega-dropdown) documents it.

jelbourn commented:
The article @joejordanbrown posted makes a pretty good point; we should consider adding a feature like that to the cdk and using in the menu as a follow-up.

crisbeto commented:
Makes sense, I'll add it to my backlog.

jelbourn commented on src/lib/menu/menu-trigger.ts:
```diff
@@ -255,48 +274,108 @@ export class MdMenuTrigger implements AfterViewInit, OnDestroy {
+    let offsetY = 0;
 
-    if (!this.menu.overlapTrigger) {
+    if (this.triggersSubmenu) {
```
Can you add some comments around how the positioning is different when triggering a submenu?

crisbeto commented on src/lib/menu/menu-item.ts:
```diff
@@ -26,15 +27,22 @@ export const _MdMenuItemMixinBase = mixinDisabled(MdMenuItemBase);
-export class MdMenuItem extends _MdMenuItemMixinBase implements Focusable, CanDisable {
+export class MdMenuItem extends _MdMenuItemMixinBase implements Focusable, CanDisable, OnDestroy {
+  /** Stream that emits when the menu item is hovered. */
+  hover: Subject<MdMenuItem> = new Subject();
```
It is mainly to facilitate the nested menu, but I can see it being useful in other cases as well.

crisbeto commented:
I'll add some UX improvements in a follow-up. This PR is mainly for figuring out the positioning, accessibility and opening/closing logic.

crisbeto commented:
Addressed the feedback @jelbourn.

Splaktar commented:
I agree with @m98 that the animation speed should be slowed down from what the demo has. It should match what the first level menu uses.

xuanshenbo commented:
wrt adding an indicator, could be related to #5067

kara commented on src/lib/menu/menu-trigger.ts:
```diff
@@ -260,48 +282,110 @@ export class MdMenuTrigger implements AfterViewInit, OnDestroy {
+    this._parentMenuSubscriptions.push(this._parentMenu.close.subscribe(() => {
+      this.menu.close.emit();
+    }));
+  }
```
Is it at all possible to avoid having a list of subscriptions? I think the logic might be neater / easier to follow if we merge the streams that result in the menu closing into one subscription, rather than having a few separate subscriptions that do the same thing.

```ts
get _menuClosingActions() {
   return merge(
      this._overlayRef.backdropClick,
      this._parentMenu.close,
      this._parentMenu.hover().filter(active => active !== this._menuItemInstance);
}
```

Then instead of only subscribing to the backdrop when the menu opens, we subscribe to the menu closing events when the menu opens, kind of like the autocomplete.

```ts
_subscribeToMenuClose() {
   this._closingSubscription = this._menuClosingActions.subscribe(() => this.menu.close.emit());
}
```

Ultimately, we'd still be adding another subscription onInit for the parent menu hovering (to open the menu when the active item is this menu item), but at least we wouldn't be adding an array of subscriptions that we'd need to manage.

crisbeto replied:
I did it initially, because they subscribe at different times in the lifecycle. Looking at it again, they don't really need to do that. Refactored.

crisbeto commented:
Rebased and addressed the feedback. Also I agree on the UX improvements, but I'll add them in a follow-up. The main purpose of this PR was to sort out the opening/closing logic, accessibility and positioning. Can you take another look @kara?

kara commented on src/lib/menu/menu-trigger.ts:
```diff
@@ -285,48 +306,96 @@ export class MdMenuTrigger implements AfterViewInit, OnDestroy {
+  /** Returns a stream that emits whenever an action that should close the menu occurs. */
+  private _menuClosingActions() {
+    const backdrop = this._overlayRef!.backdropClick();
+    const parentClose = !this._parentMenu ? observableOf(null) : this._parentMenu.close;
```
Nit: I'd switch the logic here and on line 364 because it's harder to scan "if not, else" statements (the else becomes "if not not").

```ts
this._parentMenu ?  this._parentMenu.close : observableOf(null);
```

crisbeto commented:
Good point WRT to closing all of the menu on click @kara. I've addressed it together with the other comment.

kara commented:
Note: Addition of `isSubmenu` to the interface is a breaking change and will need to be updated internally before merge.

crisbeto commented:
I suppose the same can be said about the `direction` and the signature change of `close`. Should I scope these changes to the `MdMenu` since they're there to facilitate the nested menus?

kara commented:
@crisbeto No changes needed on your end; just a note to myself so I don't accidentally merge prematurely.

kara commented:
@crisbeto Actually, chatted with @jelbourn and making `isSubmenu` optional would be even better.

crisbeto commented:
Done.
